### DISCUSSION TITLE
osd: 6 osds: 0 up, 6 in

### DISCUSSION THREAD

**Discussion by bcspragu** [rook/rook#12201] [Q&A]
Hi there,

I have a Rook/Ceph cluster, which I royally broke when my `etcd` cluster went unhealthy and all nodes temporarily went down. I've since gotten the cluster back up, but Rook is in rough shape.

Originally, the OSDs were all crash-looping with auth-related issues, but I fixed that with [these instructions](https://github.com/rook/rook/issues/4238#issuecomment-664627282). Now, all the OSDs are running, but not actually functioning. I found [this issue](https://github.com/rook/rook/issues/8329) which seemed to have the same behavior, but [the linked documentation](https://github.com/rook/rook/pull/7154/files) was never merged.

Basically, I think all the data of the cluster _should_ be intact, I'm just lost as for how to bring it back up. Some diagnostic things:

<details>
<summary>ceph status</summary>

```
  cluster:
    id:     2b10c2f1-3c72-4824-adda-1c8a7de6c26f
    health: HEALTH_WARN
            mons are allowing insecure global_id reclaim
            1 MDSs report slow metadata IOs
            Reduced data availability: 5 pgs inactive
            4 slow ops, oldest one blocked for 7423 sec, mon.b has slow ops

  services:
    mon: 3 daemons, quorum c,b,e (age 3m)
    mgr: a(active, since 12h), standbys: b
    mds: 1/1 daemons up, 1 standby
    osd: 6 osds: 0 up, 6 in (since 103m)

  data:
    volumes: 1/1 healthy
    pools:   5 pools, 5 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     100.000% pgs unknown
             5 unknown
``` 

</details>

<details>
<summary>ceph health detail</summary>

```
HEALTH_WARN mons are allowing insecure global_id reclaim; 1 MDSs report slow metadata IOs; Reduced data availability: 5 pgs inactive; 4 slow ops, oldest one blocked for 7483 sec, mon.b has slow ops
[WRN] AUTH_INSECURE_GLOBAL_ID_RECLAIM_ALLOWED: mons are allowing insecure global_id reclaim
    mon.c has auth_allow_insecure_global_id_reclaim set to true
    mon.b has auth_allow_insecure_global_id_reclaim set to true
    mon.e has auth_allow_insecure_global_id_reclaim set to true
[WRN] MDS_SLOW_METADATA_IO: 1 MDSs report slow metadata IOs
    mds.fs-ec-b(mds.0): 31 slow metadata IOs are blocked > 30 secs, oldest blocked for 46156 secs
[WRN] PG_AVAILABILITY: Reduced data availability: 5 pgs inactive
    pg 1.0 is stuck inactive for 12h, current state unknown, last acting []
    pg 2.0 is stuck inactive for 12h, current state unknown, last acting []
    pg 3.0 is stuck inactive for 12h, current state unknown, last acting []
    pg 4.0 is stuck inactive for 12h, current state unknown, last acting []
    pg 5.0 is stuck inactive for 12h, current state unknown, last acting []
[WRN] SLOW_OPS: 4 slow ops, oldest one blocked for 7483 sec, mon.b has slow ops

``` 

</details>

<details>
<summary>High-level view of rook-ceph namespace</summary>

![2023-05-07_11-46-51](https://user-images.githubusercontent.com/2402010/236696754-b29f1ee1-3896-4e3a-98cf-f1a6a7d05b20.png)

</details>

<details>
<summary>Recent manager logs</summary>

```
debug 2023-05-07T18:47:26.099+0000 7fbbd6220700  0 [dashboard INFO root] server: ssl=yes host=:: port=8443                                                                 
debug 2023-05-07T18:47:26.099+0000 7fbbd6220700  0 [dashboard INFO root] Config not ready to serve, waiting: no certificate configured                                     
debug 2023-05-07T18:47:26.183+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.4 v2:10.244.2.31:6800/218582277; not ready for session (expect reconnect) 
debug 2023-05-07T18:47:26.183+0000 7fbcb9047700  1 mgr finish mon failed to return metadata for osd.4: (2) No such file or directory                                       
debug 2023-05-07T18:47:26.223+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.5 v2:10.244.1.35:6800/578360046; not ready for session (expect reconnect) 
debug 2023-05-07T18:47:26.223+0000 7fbcb8846700  1 mgr finish mon failed to return metadata for osd.5: (2) No such file or directory                                       
debug 2023-05-07T18:47:26.247+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.1 v2:10.244.2.30:6800/3616986151; not ready for session (expect reconnect)
debug 2023-05-07T18:47:26.247+0000 7fbcb9047700  1 mgr finish mon failed to return metadata for osd.1: (2) No such file or directory                                       
debug 2023-05-07T18:47:26.295+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.0 v2:10.244.0.34:6800/1685645835; not ready for session (expect reconnect)
debug 2023-05-07T18:47:26.295+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.3 v2:10.244.0.33:6800/2448204346; not ready for session (expect reconnect)
debug 2023-05-07T18:47:26.295+0000 7fbcb8846700  1 mgr finish mon failed to return metadata for osd.0: (2) No such file or directory                                       
debug 2023-05-07T18:47:26.295+0000 7fbcb9047700  1 mgr finish mon failed to return metadata for osd.3: (2) No such file or directory                                       
debug 2023-05-07T18:47:26.355+0000 7fbbf7937700  2 mgr.server handle_open ignoring open from osd.2 v2:10.244.1.34:6800/3834228061; not ready for session (expect reconnect)
debug 2023-05-07T18:47:26.355+0000 7fbcb8846700  1 mgr finish mon failed to return metadata for osd.2: (2) No such file or directory                                       
```
</details>

<details>
<summary>Recent operator log</summary>

```
clusterdisruption-controller: all "host" failure domains: [talos-45i-j3e talos-b4h-ac9 talos-cdl-ujy]. osd is down in failure domain: "". active node drains: false. pg health: "cluster is not fully clean. PGs: [{StateName:unknown Count:5}]"                           
```
</details>

<details>
<summary>Recent operator log</summary>

```
clusterdisruption-controller: all "host" failure domains: [talos-45i-j3e talos-b4h-ac9 talos-cdl-ujy]. osd is down in failure domain: "". active node drains: false. pg health: "cluster is not fully clean. PGs: [{StateName:unknown Count:5}]"                           
```
</details>

**Comment by travisn** [comment]
#7154 would be a separate issue/solution. It assumes the mon quorum is lost, but in your case I understand that the mons are perfectly in quorum, but the OSDs are not coming back `up` properly. 

Now the OSD auth matches what you see in `osd auth ls`, right? I believe you're saying at least the OSDs came online after that was fixed.

What version of rook and ceph do you have? I'm assuming not the latest since you have this warning. Newer clusters should have that disabled. It shouldn't affect the cluster functionality, so I don't believe that's related.
`auth_allow_insecure_global_id_reclaim set to true`

You might try increasing the osd log level to 10 (or even 20) and see if you can get more details about why it's not coming back `up`. 
https://docs.ceph.com/en/latest/rados/troubleshooting/log-and-debug/

**Reply by bcspragu** [reply]
Thanks for the help Travis! Addressing your comments:

> #7154 would be a separate issue/solution. It assumes the mon quorum is lost, but in your case I understand that the mons are perfectly in quorum, but the OSDs are not coming back `up` properly.

Indeed, my understanding is that mon infra is functional in my cluster.

> Now the OSD auth matches what you see in `osd auth ls`, right? I believe you're saying at least the OSDs came online after that was fixed.

Indeed, When I run `ceph auth ls`, the osds have the keys that I had previously imported with `ceph auth import -i <key file>`. They were previously crash looping with auth issues, and now they're running.

> What version of rook and ceph do you have? I'm assuming not the latest since you have this warning. Newer clusters should have that disabled. It shouldn't affect the cluster functionality, so I don't believe that's related. `auth_allow_insecure_global_id_reclaim set to true`

It shouldn't be _too_ old of a version, I only set this cluster up ~1-2 months ago, and used the latest version then:

```
bash-4.4$ ceph version
ceph version 17.2.5 (98318ae89f1a893a6ded3a640405cdbb33e08757) quincy (stable)
```

It looks like the K8s operator is running `rook/ceph:v1.11.1`
 
> You might try increasing the osd log level to 10 (or even 20) and see if you can get more details about why it's not coming back `up`. https://docs.ceph.com/en/latest/rados/troubleshooting/log-and-debug/

Following those instructions and running `ceph tell osd.1 config set debug_osd 10`, I get an error:

```
Error ENXIO: problem getting command descriptions from osd.1
```

Not sure if that gives us any leads or not. Googling around for the error, I didn't see any obvious root cause (version mismatches, OSDs not working, etc). As for `osd.1` itself, the logs show a bunch of stats and histograms, and then:

```
│ debug 2023-05-08T23:23:26.832+0000 7fad607fa700  1 osd.1 798 tick checking mon for new map                                                                  │
│ debug 2023-05-08T23:23:57.795+0000 7fad607fa700  1 osd.1 798 tick checking mon for new map                                                                  │
│ debug 2023-05-08T23:24:27.997+0000 7fad607fa700  1 osd.1 798 tick checking mon for new map                                                                  │
│ debug 2023-05-08T23:24:58.763+0000 7fad607fa700  1 osd.1 798 tick checking mon for new map                                                                  │
```

Also not sure if that's an indication of a healthy OSD.

**Reply by bcspragu** [reply]
And for a bit more context, I think I understand at least one of my major screw-ups: Part of getting my cluster (K8s, not Rook) was [resetting the control plane nodes](https://www.talos.dev/v1.4/advanced/disaster-recovery/#preparing-control-plane-nodes) (running [Talos](https://www.talos.dev/)), which wipes `/var`, which as I understand it, is where Rook stores lots of important operating information.

So it seems like I'll need to rebuild the cluster from the OSD metadata, which [seems possible, if involved](https://docs.ceph.com/en/latest/cephfs/disaster-recovery-experts/).

**Reply by travisn** [reply]
If the host path is wiped, it certainly does cause the disaster scenarios if you're using the host-based cluster. In cloud environments where nodes are disposable, the [PVC-based cluster](https://rook.io/docs/rook/latest/CRDs/Cluster/pvc-cluster/) would be preferred so the mons and OSDs can easily move between nodes.

**Reply by bcspragu** [reply]
> In cloud environments where nodes are disposable, the [PVC-based cluster](https://rook.io/docs/rook/latest/CRDs/Cluster/pvc-cluster/) would be preferred so the mons and OSDs can easily move between nodes.

Unfortunately, in this case "the cloud" is a [Dell M1000e](https://www.dell.com/ae/business/p/poweredge-m1000e/pd) in my basement, with an extremely finite amount of storage and mostly-non-disposable nodes, so a provisioner doesn't fit nicely, as all the storage is already there and allocated to the cluster.

Is there any documentation on recovering data in this (host path wiped) scenario? Even simply being able to reassemble and extract data from the underlying storage would be sufficient to make a backup + start from scratch.

**Reply by travisn** [reply]
Got it, not much of a cloud environment for a pvc-based cluster!

There was some preliminary documentation for recovering based on the OSD data on disk found in #11895, but it was not finished. That effort is now tracked in the krew plugin with https://github.com/rook/kubectl-rook-ceph/issues/74, but that's not much help for now. Hopefully the former could help to some degree.
